src/Microsoft.ML.TensorFlow/TensorflowUtils.cs

```diff
@@ -98,6 +103,138 @@ internal static DataViewSchema GetModelSchema(IHostEnvironment env, string model
             return GetModelSchema(env, model.Session.graph);
         }
 
+        /// <summary>
+        /// Load TensorFlow model into memory.
+        /// </summary>
+        /// <param name="env">The environment to use.</param>
+        /// <param name="modelPath">The model to load.</param>
+        /// <returns></returns>
+        internal static TensorFlowModel LoadTensorFlowModel(IHostEnvironment env, string modelPath)
+        {
+            var session = GetSession(env, modelPath);
+            return new TensorFlowModel(env, session, modelPath);
+        }
+
+        internal static PrimitiveDataViewType Tf2MlNetType(TF_DataType type)
+        {
+            var mlNetType = Tf2MlNetTypeOrNull(type);
+            if (mlNetType == null)
+                throw new NotSupportedException("TensorFlow type not supported.");
+            return mlNetType;
+        }
+
+        internal static PrimitiveDataViewType Tf2MlNetTypeOrNull(TF_DataType type)
+        {
+            switch (type)
+            {
+                case TF_DataType.TF_FLOAT:
+                    return NumberDataViewType.Single;
+                case TF_DataType.DtFloatRef:
+                    return NumberDataViewType.Single;
+                case TF_DataType.TF_DOUBLE:
+                    return NumberDataViewType.Double;
+                case TF_DataType.TF_UINT8:
+                    return NumberDataViewType.Byte;
+                case TF_DataType.TF_UINT16:
+                    return NumberDataViewType.UInt16;
+                case TF_DataType.TF_UINT32:
+                    return NumberDataViewType.UInt32;
+                case TF_DataType.TF_UINT64:
+                    return NumberDataViewType.UInt64;
+                case TF_DataType.TF_INT8:
+                    return NumberDataViewType.SByte;
+                case TF_DataType.TF_INT16:
+                    return NumberDataViewType.Int16;
+                case TF_DataType.TF_INT32:
+                    return NumberDataViewType.Int32;
+                case TF_DataType.TF_INT64:
+                    return NumberDataViewType.Int64;
+                case TF_DataType.TF_BOOL:
+                    return BooleanDataViewType.Instance;
+                case TF_DataType.TF_STRING:
+                    return TextDataViewType.Instance;
+                default:
+                    return null;
+            }
+        }
+
+        internal static Session LoadTFSession(IExceptionContext ectx, byte[] modelBytes, string modelFile = null)
+        {
+            var graph = new Graph();
+            try
+            {
+                graph.Import(modelBytes, "");
+            }
+            catch (Exception ex)
+            {
+                if (!string.IsNullOrEmpty(modelFile))
+                    throw ectx.Except($"TensorFlow exception triggered while loading model from '{modelFile}'");
```

Comment:
This should have the `ex` as an inner exception. So that information doesn't get lost.

Reply:
thanks!

---
In reply to: [340221260](https://github.com/dotnet/machinelearning/pull/4408#discussion_r340221260) [](ancestors = 340221260)